### PR TITLE
[#81] Add unit tests for the underlying SSH modules

### PR DESCRIPTION
**Pull request by klappradla**
This adds some more unit tests for the underlying SSH modules. There are however some exception, where I could not really come up with a sane way of testing these things, e.g. for `Channel.loop/4` - I guess it would be better covered by having more high level tests.

In addition to that, it would be cool to define certain boundaries on which parts of the library are supposed to be "public api" and which not. As for instance, `SSHKit.SSH.Channel` may be too low-level and not really the way how people should use `SSHKit`...?
From what I've seen, the best practise to set up these boundaries is probably only to have these modules `@moduledoc false`.. What do you think? (cc @tessi & @pmeinhardt)

For these unit tests, I also changed the way we set up our `ssh_modules` in unit tests to be more Elixir-like. I found it's best practise to make use of the multiple `setup` and `context` macros ExUnit offers and combining them with tags, rather than using module variables.
I also grouped all definitions we have for when which *mocky*  `ssh` or `ssh_connection` module is used. Using tags makes it easier to keep track of which of them is supposed to cause an error, etc. without having to remember all names by heard. This also comes with the benefit, that we can change the names in the future without breaking **ALL** the unit tests 😉 

---

* [x] Documented the change if necessary
* [x] Tested this PRs change (with unit and/or functional tests)
* [x] Added this PRs change to CHANGELOG.md (in the `#master` section) if  necessary.

### CODE REVIEW

File: test/sshkit/ssh_test.exs [hunk 1]
```diff
@@ -155,48 +126,36 @@ defmodule SSHKit.SSHTest do
...
         port:        22,
         ref:         :sandbox,
-        ssh_modules: Keyword.get(@options, :ssh_modules)
+        ssh_modules: ssh_modules
       }
-      assert run(conn, "uptime") == {:error, :closed}
-    end
 
-    @options [ssh_modules: %{ssh: SSHSandboxSuccess, ssh_connection: SSHSandboxConnectionSuccess}]
-    test "sucessfully execute command on connection and return result" do
```
Comment (tessi) [inline test/sshkit/ssh_test.exs]: you remove the test to a successful connection. is there a reason for it?

Comment (tessi) [inline test/sshkit/ssh_test.exs]: you remove the test to a successful connection. is there a reason for it?

Comment (klappradla) [reply test/sshkit/ssh_test.exs]: 🤔  I cannot really remember tbh... Looking at it now, I think it was probably only an accident 🙈 

I'm going to add it back!

**Comment by klappradla** [comment]
👍  for agreeing on the public API.

Should I add a ticket for it? It would probably be cool to have @pmeinhardt's ideas on the topic as well before diving into it.

File: test/sshkit/ssh_test.exs (hunk 1 above)
Comment (klappradla) [reply test/sshkit/ssh_test.exs]: Ok, it's back now ✌️

Comment (klappradla) [reply test/sshkit/ssh_test.exs]: Ok, getting a weird behavior of the test on Travis... Will investigate and fix

Comment (klappradla) [reply test/sshkit/ssh_test.exs]: Ah, now I remember: running a command and not being able to control the remote side will make the channel listen for messages forever and / or timeout. I could not yet come up with a sane way of testing this - would be happy for suggestions though 🙏

**Comment by tessi** [comment]
Yep, let's see what @pmeinhardt or @paulodiniz think to the public/internal API thing (could be a separate issue and a separate discussion anyways)

**Comment by pmeinhardt** [comment]
## Regarding boundaries…

Here's a bunch of diagrams for a better picture of how SSHKit.ex is structured and as a reference point for the discussion maybe – ASCII-art FTW ✏️

Overall, [SSHKit is organized like this](https://github.com/bitcrowd/sshkit.ex#modules):

```
+--------------------+
| SSHKit             |
+--------------------+
|       | SSHKit.SCP |
|       +------------+
| SSHKit.SSH         |
+--------------------+
```

Dissecting the lowest layer, `SSHKit.SSH` further, we can see that it abstracts away the Erlang `:ssh` (`SSHKit.SSH.Connection`) and `:ssh_connection` (`SSHKit.SSH.Channel`) modules:

```
+--------------------------------------------+
| SSHKit.SSH                                 |
+-----------------------+--------------------+
| SSHKit.SSH.Connection | SSHKit.SSH.Channel |
+-----------------------+--------------------+
| :ssh                  | :ssh_connection    |
+-----------------------+--------------------+
```

## Regarding official/internal…

When I started out with SSHKit, I first created a draft/goal for what I wanted the final, top-level `SSHKit` API to look like. This is the main reason why you'd use SSHKit and definitely public.

To me, a secondary contribution of SSHKit are the `SSHKit.SSH` and the `SSHKit.SCP` modules. My hope when writing these was that if you're writing Elixir code and you have anything to do with SSH or SCP, these abstractions will help you be more productive even if the main SSHKit API is too high-level for you.

Everything inside of/below these, I would consider internal/private — in particular the modules `SSHKit.SSH.{Connection, Channel}` and `SSHKit.SCP.{Command, Download, Upload}`.

---

Hope this helps. These are just my two cents… and definitely not the only possible point of view 😃

**Comment by tessi** [comment]
@klappradla regarding the failing success-case-test: I currently have no idea how to make it not timeout in a nice way. Let's create an issue for this and move on with this PR. It improves the codebase and I'd like to merge it to be able to release a new version.

**Comment by klappradla** [comment]
Ok, moved the discussion on module boundaries and public API over to https://github.com/bitcrowd/sshkit.ex/issues/90 - so that we don't forget about it *(I also already included some of @pmeinhardt's artworks there)*

**Pull request merged**